File: src/test/test_mesh.f90
program test_mesh
  use control
  use mesh

  implicit none

  ! initialize types
  integer :: number_cells_test, mMap_test(8)
  double precision :: dx_test(8)
  integer :: t1=1, t2=1, t3=1, testCond
  
  call initialize_control('test/reg_test_options', .true.)
  ! Define problem parameters
  fine_mesh = [2, 4, 2]
  material_map = [1, 2, 3]
  course_mesh = [0.0, 1.0, 2.0, 3.0]
  boundary_type = [0.0, 0.0]
  
  ! Make the mesh
  call create_mesh()
  
  number_cells_test = 8
  dx_test = [0.5, 0.5, 0.25, 0.25, 0.25, 0.25, 0.5, 0.5]
  mMap_test = [1, 1, 2, 2, 2, 2, 3, 3]

  ! Test number of cells
  t1 = testCond(number_cells_test == number_cells)
  
  ! test dx
  t2 = testCond(norm2(dx-dx_test) < 1e-7)
  
  ! test material map
  t3 = testCond(ALL(mMap == mMap_test))
  
  ! Print appropriate output statements
  if (t1 == 0) then
    print *, 'mesh: number_cells failed'
  else if (t2 == 0) then
    print *, 'mesh: dx failed'
  else if (t3 == 0) then
    print *, 'mesh: mMap failed'
  else
    print *, 'all tests passed for mesh'
  end if

  call finalize_mesh()
  call finalize_control()

end program test_mesh

integer function testCond(condition)
  logical, intent(in) :: condition
  if (condition) then
    write(*,"(A)",advance="no") '.'
    testCond = 1
  else
    write(*,"(A)",advance="no") 'F'
    testCond = 0
  end if

end function testCond

File: src/test/test_mesh.py
import sys
sys.path.append('../')

import unittest
import pydgm
import numpy as np


class TestMESH(unittest.TestCase):

    def test_mesh_create_mesh_1D(self):
        ''' 
        Test the mesh initialization for 1D
        '''

        # Set the variables for the test
        pydgm.control.spatial_dimension = 1
        pydgm.control.fine_mesh_x = [2, 4, 2]
        pydgm.control.coarse_mesh_x = [0.0, 1.0, 2.0, 3.0]
        pydgm.control.material_map = [1, 2, 3]

        # Initialize the dependancies
        pydgm.mesh.create_mesh()

        # Test the number of cells
        self.assertEqual(pydgm.control.number_cells, 8)

        # Test the cell size
        dx_test = [0.5, 0.5, 0.25, 0.25, 0.25, 0.25, 0.5, 0.5]
        np.testing.assert_array_equal(pydgm.mesh.dx, dx_test)

        # Test the material assignment
        mMap_test = [1, 1, 2, 2, 2, 2, 3, 3]
        np.testing.assert_array_equal(pydgm.mesh.mmap, mMap_test)

        # Test the problem width
        self.assertEqual(pydgm.mesh.width_x, 3.0)

    def test_mesh_create_mesh_2D(self):
        ''' 
        Test the mesh initialization for 2D
        '''

        # Set the variables for the test
        pydgm.control.spatial_dimension = 2
        pydgm.control.fine_mesh_x = [5, 5, 3]
        pydgm.control.fine_mesh_y = [4, 4, 2]
        pydgm.control.coarse_mesh_x = [0.0, 21.42, 42.84, 64.26]
        pydgm.control.coarse_mesh_y = [0.0, 21.42, 42.84, 64.26]
        pydgm.control.material_map = [2, 4, 5,
                                      4, 2, 5,
                                      5, 5, 5]

        # Initialize the dependancies
        pydgm.mesh.create_mesh()

        # Test the number of cells
        self.assertEqual(pydgm.control.number_cells, 130)
        self.assertEqual(pydgm.control.number_cells_x, 13)
        self.assertEqual(pydgm.control.number_cells_y, 10)

        # Test the cell size
        dx_test = [4.284, 4.284, 4.284, 4.284, 4.284, 4.284, 4.284, 4.284, 4.284, 4.284, 7.14, 7.14, 7.14]
        dy_test = [5.355, 5.355, 5.355, 5.355, 5.355, 5.355, 5.355, 5.355, 10.71, 10.71, 10.71]
        np.testing.assert_array_almost_equal(pydgm.mesh.dx, dx_test, 12)
        np.testing.assert_array_almost_equal(pydgm.mesh.dx, dx_test, 12)

        # Test the material assignment
        mMap_test = [2, 2, 2, 2, 2, 4, 4, 4, 4, 4, 5, 5, 5,
                     2, 2, 2, 2, 2, 4, 4, 4, 4, 4, 5, 5, 5,
                     2, 2, 2, 2, 2, 4, 4, 4, 4, 4, 5, 5, 5,
                     2, 2, 2, 2, 2, 4, 4, 4, 4, 4, 5, 5, 5,
                     4, 4, 4, 4, 4, 2, 2, 2, 2, 2, 5, 5, 5,
                     4, 4, 4, 4, 4, 2, 2, 2, 2, 2, 5, 5, 5,
                     4, 4, 4, 4, 4, 2, 2, 2, 2, 2, 5, 5, 5,
                     4, 4, 4, 4, 4, 2, 2, 2, 2, 2, 5, 5, 5,
                     5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5,
                     5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5]
        np.testing.assert_array_equal(pydgm.mesh.mmap, mMap_test)

        # Test the problem width
        self.assertEqual(pydgm.mesh.width_x, 64.26)
        self.assertEqual(pydgm.mesh.width_y, 64.26)

    def tearDown(self):
        pydgm.mesh.finalize_mesh()
        pydgm.control.finalize_control()


if __name__ == '__main__':

    unittest.main()
